feat: создать каркас приложения

NekhoroshkinaElena opened the pull request.

NekhoroshkinaElena committed:
feat: создать каркас приложения

NekhoroshkinaElena committed:
fix: исправить ошибку с аннотациями в User.java

FanaticFiz commented on src/main/java/ru/practicum/shareit/booking/model/Status.java:
```diff
@@ -0,0 +1,8 @@
+package ru.practicum.shareit.booking.model;
```
`pom` файл не менялся.
Напишу коммент в произвольном месте...

1. Хорошей практикой будет указать кодировку 
`<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>`

2. Я советую всем соблюдать некий порядок при объявлении зависимостей.
Сначала "главный" `spring` все его зависимости и всё о нём. 
Потом по степени "тяжеловестности" остальные зависимости

ну и у меня в проекте, например, последними идёт еще одна секция наши собственные библиотеки/модули/контракты.

FanaticFiz commented on src/main/java/ru/practicum/shareit/booking/model/Status.java:
```diff
@@ -0,0 +1,8 @@
+package ru.practicum.shareit.booking.model;
+
+public enum Status {
```
Будь аккуратна с таким общим названием: `Status`, бывает так что оно пересекается с такими же из других библиотек и это может съесть у тебя некоторое время...
Например, можно случайно заимпортить не то и просидеть часик в отладке не понимая что происходит! )

FanaticFiz commented on src/main/java/ru/practicum/shareit/exception/ErrorHandler.java:
```diff
@@ -0,0 +1,16 @@
+
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public ErrorResponse handleFilmNotFoundException(final ValidationException e) {
```
Мы не говорим `handleFilmNotFoundException` надо обще говорить что что то не найдено: `handleNotFoundException`.

Мы не создаём классы исключений под каждую сущность...
Создаётся один общий класс исключений.
Бросая его, ты в сообщении будешь писать, что именно произошло.
Это первое...

Второе.
Почему на `ValidationException` ответ `HttpStatus.NOT_FOUND` ? 
Это неверно. На ошибки валидации нужно отвечать `BadRequest`
А `NotFoundException` я у тебя и не вижу!

FanaticFiz commented on src/main/java/ru/practicum/shareit/booking/model/Status.java:
```diff
@@ -0,0 +1,8 @@
+    APPROVED,
+    REJECTED,
+    CANCELED
+}
```
И про файл настроек `application.properties`, так как он не менялся, тоже тут напишу...

Я буду топить за то чтобы ты сразу привыкала к написанию файла свойств для приложения в формате `yaml -> application.yml`
Мне трудно сейчас представить проект в котором до сих пор на него не пересели ))

Соответственно - попрошу тебя этому следовать.

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/ItemMapper.java:
```diff
@@ -0,0 +1,15 @@
+
+public class ItemMapper {
+
+    public static ItemDto toItemDto(Item item) {
```
Мапперы должны давать методы для конвертирования в обе стороны. 
У тебя сейчас только в одну.

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,33 @@
+    private String description;
+    @NotNull
+    private Boolean available;
+    private String owner;
```
Пользователь и запрос должны быть представлены сущностями!

```
private User owner;
private ItemRequest request;
```

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,13 @@
+package ru.practicum.shareit.user.dto;
+
+public class UserDto {
```
Похоже этот класс недооформлен и не используется!

Напомню что dto принимаются контроллерами.
Эти классы обычно содержат аннотации для валидации.

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,44 @@
+    }
+
+    @PostMapping
+    public User saveUser(@Valid @RequestBody User user) {
```
В контроллерах используются dto. Оставлю комментарий и тут...

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/UserRepositoryInMemory.java:
```diff
@@ -0,0 +1,64 @@
+
+    public User addUser(User user) {
+        for (User u : users.values()) {
+            if (u.getEmail().equals(user.getEmail())) {
```
Обрати внимание, эта проверка у тебя написана несколько раз в этом классе.
Вынеси её в отдельный метод - старайся писать код только один раз! 

```
            if (u.getEmail().equals(user.getEmail())) {
                log.error("Пользователь с таким email уже существует");
                throw new RuntimeException("Пользователь с таким email уже существует");
            }
```

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -0,0 +1,27 @@
+    private long id;
+    private String name;
+    @NotBlank
+    @Email
```
Аннотации по валидации должны переехать в dto

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/service/UserService.java:
```diff
@@ -0,0 +1,17 @@
+
+import java.util.List;
+
+public interface UserService {
```
Что касается интерфейса и методов в нем.
- проблема всех названий - не добавляй `User`! Методы не висят в воздухе, они находятся в контексте класса.
Поэтому это лишнее. Просто `add`, `update`

Когда ты тоже самое проделаешь и со вторым интерфейсом, то ты наверняка заметишь их сходство.
Подумай можно ли и с этим что то сделать?

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public User getUser(long id) {
+        return userRepository.getUser(id);
```
Нет проверки на то что такого пользователя нет... 
Поэтому и не создан еще соответствующий класс исключений! Исправься!

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public User userUpdate(User user, long id) {
+        return userRepository.userUpdate(user, id);
```
Кстати вон ту проверку на то что пользователь уже существует надо вынести в сервис.

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,22 @@
+    private String name;
+    private String description;
+    private Boolean available;
+
```
не хватает `private Long requestId;` ?

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/ItemRepositoryInMemory.java:
```diff
@@ -0,0 +1,115 @@
+    private final UserRepositoryInMemory repository;
+    private final Map<Long, Item> items = new HashMap<>();
+    private Map<Long, ItemDto> itemsDto = new HashMap<>();
+    private final Map<Long, Long> users = new HashMap<>(); //key - id вещи, values - id пользователя
```
Тут твориться чтот странное... почему тут существуют пользователи?
Они же в своём хранилище...

Всё дело в том что ты пытаешься всё делать в хранилищах! А сервисы у тебя пусты...
Хранилище должно только отдавать сущности.
А логика приложения в сервисах.

А сервисы имеют полное право оперировать разными хранилищами
ItemService будет включать
```
	UserRepository userRepository;
	ItemRepository itemRepository;
```
Что даст тебе возможность работать с пользователями из их родного репозитория...
Как то так ))

NekhoroshkinaElena committed:
fix: изменить название класса enum Status на BookingStatus, во избежание пересечения с такими же названиями классов из других библиотек
fix: добавить класс NotFoundException и исправить ответ на ValidationException c NOT_FOUND на BAD_REQUEST
fix: добавить метод для конвертирования из Dto в ItemMapper и UserMapper
fix: изменить тип полей owner и request в классе Item
fix: перенести аннотации для валидации в dto классы
fix: изменить название всех методов
feat: добавить проверку на существование пользователя в UserServiceImpl
fix: добавить в ItemDto поле requestId

NekhoroshkinaElena committed:
fix: исправить перенос

NekhoroshkinaElena committed:
fix: убрать дублирование кода в ItemServiceImpl

NekhoroshkinaElena committed:
refactor: изменить application.properties на application.yml

NekhoroshkinaElena committed:
fix: убрать лишние аннотации из классов ItemRepositoryInMemory и UserRepositoryInMemory

NekhoroshkinaElena commented:
Добрый день! Большое спасибо за такие подробные комментарии, действительно, многое стало понятно! постаралась устранить ошибки:)

FanaticFiz commented on src/main/java/ru/practicum/shareit/exception/ErrorHandler.java:
```diff
@@ -0,0 +1,22 @@
+    public ErrorResponse handleNotFoundException(final NotFoundException e) {
+        return new ErrorResponse(e.getMessage());
+    }
+}
```
Ты не предусматриваешь ситуации, при которой есть повод кидать HttpStatus.CONFLICT` ?
Я как бы намекаю на создание пользователей с одним и тем же email.

Будет отдельный класс исключения с соответствующим названием.

FanaticFiz replied:
А так же не хватает 

```
@ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
	public ErrorResponse handleThrowable(final Throwable e) {
```

на тот случай когда приложение будет падать неожиданно для нас.
Такое случается чаще чем хочется ! ))

FanaticFiz commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,24 @@
+spring:
+  jpa:
+    hibernate.ddl-auto: none
```
Не дожала же перевести свойства в формат yml.
Ну например: 
`hibernate.ddl-auto: none`

```
    hibernate:
      ddl-auto: none
```

Остальное найди в инете если так просто не получается.

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,23 @@
+@Setter
+public class UserDto {
+    private long id;
+    private String name;
```
Не дожала и валидацию! 
Разве мы разрешаем создать пользователя с пустым `name` ?

Так что как минимум @NotBlank нужен для name

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,48 @@
+
+    @PostMapping
+    public UserDto save(@Valid @RequestBody UserDto userDto) {
+        User user = UserMapper.toUser(userDto);
```
Елена, нет не так...

`Контроллер и сервис` Передают между собой `UserDto` модели
`Сервис и репозиторий` Передают между собой `User` модели

Таким образом, маппер используется в сервисе. Поскольку это он работает на две стороны.

```
Controller <------> Service <------> Repository
                    mappers
```

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,48 @@
+
+    @GetMapping
+    public List<UserDto> getAll() {
+        return userService.getAll().stream().map(UserMapper::toUserDto).collect(Collectors.toList());
```
Ну и соответственно расположение мапперов не там, пронизывает всю твою работу.
Тут должен быть только `eturn userService.getAll()`

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,58 @@
+        for (User u : userRepository.getAll()) {
+            if (u.getEmail().equals(user.getEmail())) {
+                log.error("Пользователь с таким email уже существует");
+                throw new RuntimeException("Пользователь с таким email уже существует");
```
Аааа вот и это место... видишь, тут ты поленилась создать свой класс исключения ))

Ну ничего... я же тут как раз для этого! Я тебя заставлю! ))))

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+        return itemRepositoryInMemory.search(text);
+    }
+
+    public void notFoundUser(long userId) {
```
Метод который что то `throw` называй честно, так чтобы было понятно, что в методе нас ждет...

`throwIfUserNotFound`

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+
+    public void notFoundUser(long userId) {
+        if (userRepositoryInMemory.getById(userId) == null) {
+            log.error("пользователя не существует");
```
Давай в лог больше полезной информации!
Не просто `пользователя не существует` а положи в лог userId по которому мы пользователя искали!

FanaticFiz reviewed (changes requested):
Финальный комментарий на Я.Практикуме

NekhoroshkinaElena committed:
feat: создать класс AlreadyExistsException
fix: внести правки в application.yml
fix: исправить название метода в классе ItemServiceImpl
feat: добавить проверку для поля name в классе UserDto
refactor: убрать дублирование в классе UserServiceImpl
feat: добавить дополнительную информацию в log

NekhoroshkinaElena committed:
fix: добавлен пробел

NekhoroshkinaElena committed:
refactor: перенести маппер из ItemController и UserController в ItemServiceImpl и UserServiceImpl

NekhoroshkinaElena commented:
Вновь постаралась исправить ошибки:)
